src/amber/router/route_set.cr

```diff
@@ -0,0 +1,285 @@
+        @branches.size == 0
+      end
+
+      def routes? : Bool
+        @branches.any?
+      end
+
+      # Recursively count the number of discrete paths remaining in the tree.
+      def size
+        return 1 if leaf?
+
+        @branches.reduce 0 do |count, branch|
+          count += branch.size
+        end
+      end
+
+      # Recursively descend to find the attached application route.
+      # Weakness: assumes only one path remains in the tree.
+      def route
+        return @route if leaf?
+        @branches.first.route
+      end
+
+      # Recursively _prunes_ the route tree by matching segments
+      # against path segment strings.
+      #
+      # A destructive breadth first search.
+      #
+      # return true if any routes matched.
+      def select_routes!(path : String) : Bool
+        first_segment, remaining_path = split_path path
+
+        reverse_match = false
+
+        case
+        when root?
+          # select all branches that match the full path
+          remaining_path = path
+        when fixed?
+          # select branches only if this segment matches
+          segment_match? first_segment
+        when variable?
+          # always match
+        when glob?
+          reverse_match = true
+        end
+
+        if reverse_match
+          match, _ = reverse_select_routes! path
+          return match
+        else
+          @branches.select! do |subtree|
+            subtree.select_routes! remaining_path
+          end
+        end
+
+        @branches.any? || (leaf? && remaining_path == "")
+      end
+
+      # Recursively matches the right hand side of a glob segment.
+      # Allows for routes like /a/b/*/d/e and /a/b/*/f/g to coexist.
+      # This is a modified version of a destructive depth first search.
+      #
+      # Importantly, each subtree must pass back up the remaining part
+      # of the path so it can be matched against the parent, so this
+      # method somewhat awkwardly returns:
+      #
+      #   Tuple(subtree_match : Bool, path_for_trunk_to_match : String)
+      #
+      def reverse_select_routes!(path : String) : Tuple(Bool, String)
+        remnant_path = ""
+        was_leaf = leaf?
+
+        @branches.select! do |subtree|
+          match, remaining_path = subtree.reverse_select_routes! path
+          if match
+
+            if remnant_path != ""
+              raise "warning: overwriting remnant path"
+            end
+
+            remnant_path = remaining_path
+          end
+        end
+
+        # If this segment started as a leaf, no remant path exists. Match against the whole path.
+        remnant_path = path if was_leaf
+
+        # If this wasn't a leaf and there are no branches left, it's not a match.
+        return Tuple(Bool, String).new(false, "") unless @branches.any? || was_leaf
+
+        # If this node is the glob, at least one subtree matched (or there are none).
+        return Tuple(Bool, String).new(true, "") if glob?
```

Comment:
there has to be a better way to do this, right?

Reply:
A tuple literal:
```crystal
return { true, "" } if glob?
```
See https://carc.in/#/r/368l

Reply:
@marksiemers thanks! I did end up figuring that side out but I didn’t find a terse way to specify a tuple return type. Is there something more readable than explicitly typing [this function](https://github.com/amberframework/router/blob/master/src/route_set.cr#L160) with a tuple class literal?

Reply:
I don't know if this creates the code style that you're looking for, but you can always be explicit about the return type (the compiler will complain if it finds a code path that returns something else - not sure about exception raising):
```crystal
# This might work?
def reverse_select_routes(path : Array(String), startpos, endpos = nil) : Tuple(Array(T), Int32)
  ...
end
```